src/cart/crt_group.c

```diff
@@ -2603,42 +2750,49 @@ crt_grp_detach(crt_group_t *attached_grp)
 
 	crt_grp_ras_fini(grp_priv);
 
-	D_RWLOCK_RDLOCK(&crt_gdata.cg_rwlock);
-	for (i = 0; i < CRT_SRV_CONTEXT_NUM; i++) {
-		ctx = crt_context_lookup(i);
-		if (ctx == NULL)
-			continue;
-		rc = crt_grp_ctx_invalid(ctx, true /* locked */);
+	if (grp_priv->gp_primary) {
+		D_RWLOCK_RDLOCK(&crt_gdata.cg_rwlock);
+		for (i = 0; i < CRT_SRV_CONTEXT_NUM; i++) {
+			ctx = crt_context_lookup(i);
+			if (ctx == NULL)
+				continue;
+			rc = crt_grp_ctx_invalid(ctx, true /* locked */);
+			if (rc != 0) {
+				D_ERROR("crt_grp_ctx_invalid failed, rc: %d.\n",
+					rc);
+				D_RWLOCK_UNLOCK(&crt_gdata.cg_rwlock);
+				D_GOTO(out, rc);
+			}
+		}
+		D_RWLOCK_UNLOCK(&crt_gdata.cg_rwlock);
+
+		rc = crt_grp_lc_destroy(grp_priv);
 		if (rc != 0) {
-			D_ERROR("crt_grp_ctx_invalid failed, rc: %d.\n", rc);
-			D_RWLOCK_UNLOCK(&crt_gdata.cg_rwlock);
+			D_ERROR("crt_grp_lc_destroy failed, rc: %d.\n", rc);
 			D_GOTO(out, rc);
 		}
-	}
-	D_RWLOCK_UNLOCK(&crt_gdata.cg_rwlock);
-
-	rc = crt_grp_lc_destroy(grp_priv);
-	if (rc != 0) {
-		D_ERROR("crt_grp_lc_destroy failed, rc: %d.\n", rc);
-		D_GOTO(out, rc);
-	}
 
-	if (grp_priv == grp_gdata->gg_srv_pri_grp) {
-		crt_grp_priv_decref(grp_priv);
-		grp_gdata->gg_srv_pri_grp = NULL;
-		D_GOTO(out, rc);
-	}
+		if (grp_priv == grp_gdata->gg_srv_pri_grp) {
+			crt_grp_priv_decref(grp_priv);
+			grp_gdata->gg_srv_pri_grp = NULL;
+			D_GOTO(out, rc);
+		}
 
-	/* remove from gg_srv_grps_attached */
-	D_RWLOCK_WRLOCK(&grp_gdata->gg_rwlock);
-	d_list_for_each_entry(grp_priv_tmp, &grp_gdata->gg_srv_grps_attached,
-			      gp_link) {
-		if (crt_grp_id_identical(attached_grp->cg_grpid,
-					 grp_priv_tmp->gp_pub.cg_grpid)) {
-			found = true;
-			break;
+		/* remove from gg_srv_grps_attached */
+		D_RWLOCK_WRLOCK(&grp_gdata->gg_rwlock);
```

Comment:
This lock is released on line 2798 but in an outer scope.   Need to check if we are a primary group before releasing it?

Reply:
good catch